keysync_initiator_test.go

```diff
@@ -0,0 +1,158 @@
+package nitriding
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+)
+
+func TestRequestNonce(t *testing.T) {
+	expNonce := nonce{
+		0x14, 0x56, 0x82, 0x13, 0x1f, 0xff, 0x9c, 0xf7, 0xeb, 0xb6,
+		0x9e, 0x7b, 0xea, 0x29, 0x16, 0x49, 0xeb, 0x03, 0xa2, 0x47,
+	}
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, expNonce.B64())
+	}))
+	defer srv.Close()
+
+	retNonce, err := requestNonce(srv.URL)
+	if err != nil {
+		t.Fatalf("Failed to request nonce: %s", err)
+	}
+	if expNonce != retNonce {
+		t.Fatal("Returned nonce not as expected.")
+	}
+}
+
+func TestRequestNonceDoS(t *testing.T) {
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		nonce1 := nonce{}
+		nonce2 := nonce{}
+		fmt.Fprintf(w, "%s%s", nonce1.B64(), nonce2.B64())
+	}))
+	defer srv.Close()
+
+	if _, err := requestNonce(srv.URL); err == nil {
+		t.Fatal("Client code should have rejected long response body but didn't.")
+	}
+}
+
+func TestRequestAttDoc(t *testing.T) {
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintln(w, "foobar")
+	}))
+	defer srv.Close()
+
+	_, err := requestAttDoc(srv.URL, []byte{})
+	if err == nil {
+		t.Fatal("Client code should have rejected non-Base64 data but didn't.")
+	}
+}
+
+func TestRequestAttDocDoS(t *testing.T) {
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		maxReadLen := base64.StdEncoding.EncodedLen(maxAttDocLen)
+		buf := make([]byte, maxReadLen+1)
+		fmt.Fprintln(w, buf)
+	}))
+	defer srv.Close()
+
+	if _, err := requestAttDoc(srv.URL, []byte{}); err == nil {
+		t.Fatal("Client code should have rejected long response body but didn't.")
+	}
+}
+
+func TestProcessAttDoc(t *testing.T) {
+	// The idea of this fairly involved test is to make sure that processAttDoc
+	// can decrypt and recover the key material that's part of an attestation
+	// document.  To that end, we are using a hard-coded attestation document,
+	// NaCl key pair, and the PCR values of the attestation document.
+	pubKey := [32]byte{
+		213, 156, 108, 34, 179, 183, 69, 26, 209, 218, 58, 186, 9, 32, 237,
+		253, 46, 80, 36, 200, 169, 239, 97, 200, 17, 188, 203, 99, 151, 40,
+		10, 113,
+	}
+	privKey := [32]byte{
+		74, 137, 121, 11, 209, 38, 48, 48, 167, 157, 184, 58, 2, 110, 9, 204,
+		174, 148, 243, 154, 191, 74, 118, 90, 11, 240, 246, 131, 187, 157,
+		157, 25,
+	}
+	ourNonce := nonce{}
+	boxKey := boxKey{
+		pubKey:  &pubKey,
+		privKey: &privKey,
+	}
+
+	// Mock getPCRValues and make it return the PCR values that are in
+	// responderAttDoc.
+	null := make([]byte, 48) // An empty PCR value.
+	getPCRValues = func() (map[uint][]byte, error) {
```

Comment:
It's a little odd that the nonce is zero, the PCR registers are not (I know, we can't generate a signed document like that) and those values are here instead of the file with the actual cert. Would it make sense to move this func into keysync_shared_test.go so everything is in one place?

What about extracting the map in code directly from rawAttDoc? That would be reasonably to code locally, and It's tedious to review updates to this static data anyway.

Reply:
Yes, fixed in 237b307.